src/Control/Monad/Logic/Sequence/Internal.hs

```diff
@@ -198,44 +251,80 @@ instance Monad m => Applicative (SeqT m) where
 instance Monad m => Alternative (SeqT m) where
   {-# INLINE empty #-}
   {-# INLINEABLE (<|>) #-}
-  {-# SPECIALIZE INLINE (<|>) :: Seq a -> Seq a -> Seq a #-}
   empty = SeqT S.empty
-  m <|> n = fromView (altView m n)
+  (<|>) = alt
 
-altView :: Monad m => SeqT m a -> SeqT m a -> m (View m a)
-altView (toView -> m) n = m >>= \x -> case x of
-  Empty -> toView n
-  h :< t -> return (h :< cat t n)
-    where cat (SeqT l) (SeqT r) = SeqT (l S.>< r)
-{-# INLINE altView #-}
+alt :: Monad m => SeqT m a -> SeqT m a -> SeqT m a
+alt a b = unstream (alt_s (stream a) (stream b))
+{-# INLINE[3] alt #-}
```

Comment:
I'm quite concerned that this approach may cause asymptotic performance trouble in the face of lots of `<|>` (associated various ways, and especially recursive).

Reply:
Yeah, I'm curious about that case too. My simple benchmark does an iterative deepening search. And then I convert the generated values to a streamly stream. That conversion can either be done using something like `observeAll` or `msplit` with `unfoldrM`. It's the program where I originally discovered the quadratic slowdown in `LogicT`. So I've been using that to get a sense of whether our `<|>` implementation is dealing well with associativity.

For the version here there's no runtime difference between using `msplit` vs. `observeAll` for `SeqT`. Have you looked in `bench`? Do any of those cover the case you are imagining? It's a little bit hard to get a graph from those benchmarks so I haven't been running those between changes (also they take forever to run). But I would like to be able to answer question like this using benchmarks as a sort of ground truth on runtime.